src/shared/copy.c

```diff
@@ -795,6 +796,66 @@ static int prepare_nocow(int fdf, const char *from, int fdt, unsigned *chattr_ma
         return 0;
 }
 
+/* Copies fs-verity status.  May re-open fdt to do its job. */
+static int copy_fs_verity(int fdf, int *fdt) {
+        int r;
+
+        assert(fdf >= 0);
+        assert(fdt);
+        assert(*fdt >= 0);
+
+        r = fd_verify_regular(fdf);
+        if (r < 0)
+                return r;
+
+        struct fsverity_descriptor desc = {};
+        struct fsverity_read_metadata_arg read_arg = {
+                .metadata_type = FS_VERITY_METADATA_TYPE_DESCRIPTOR,
+                .buf_ptr = (uintptr_t) &desc,
+                .length = sizeof(desc),
+        };
+
+        r = ioctl(fdf, FS_IOC_READ_VERITY_METADATA, &read_arg);
+        if (r < 0) {
+                /* ENODATA means that the file doesn't have fs-verity,
+                 * so the correct thing to do is to do nothing at all.
+                 */
+                if (errno == ENODATA)
+                        return 0;
+                return log_error_errno(errno, "Failed to read fs-verity metadata from source file: %m");
+        }
+
+        /* Make sure that the descriptor is completely initialized */
+        assert(r == (int) sizeof desc);
+
+        /* Okay.  We're doing this now.  We need to re-open fdt as read-only because
+         * we can't enable fs-verity while writable file descriptors are outstanding.
+         */
+        r = fd_verify_regular(*fdt);
+        if (r < 0)
+                return r;
+
+        _cleanup_close_ int reopened_fd = -EBADF;
+        r = fd_reopen_condition(*fdt, O_RDONLY|O_CLOEXEC|O_NOCTTY, O_ACCMODE|O_PATH, &reopened_fd);
```

Comment:
Could you also comment here about why we need to reopen with O_RDONLY ?

Reply:
oh btw, please switch to O_ACCMODE_STRICT, see b1236ce38bb2f918c7150df36b32fcb84b1a23fb

Reply:
I won't pretend to understand why, but done :)